### PR TITLE
Read Split Checkpoint Files

### PR DESCRIPTION
**Pull request by friedmud**
Initial capability for reading split checkpoint files.

Also slightly modifies how Nemesis files are read by removing the `nemesis` option in favor of a common `pre_split` option.

### CODE REVIEW

**Comment by friedmud** [comment]
This is somewhat tied to https://github.com/libMesh/libmesh/pull/1228

Anyone trying to use this PR should use the libMesh branch in https://github.com/libMesh/libmesh/pull/1228 to split the mesh.

**Comment by friedmud** [comment]
Hmm - I don't understand the test failure - it works fine on my system...

Maybe a difference  in how libMesh is configured?

**Comment by permcody** [comment]
Do you _require_ a special branch of libmesh or even just a newer version
than the submodule? If so, make sure that's also part of your commit.

**Comment by friedmud** [comment]
Good point!  I don't think so - but let me double check...

**Comment by friedmud** [comment]
@permcody you were right - the test doesn't pass with the current libMesh sub-module.

It doesn't need that PR - but it does need the parallel checkpoint stuff that @jwpeterson got merged into libMesh a little while ago

**Comment by friedmud** [comment]
I think I'll wait until libMesh/libmesh#1228 gets merged before updating the submodule in this PR.

**Comment by friedmud** [comment]
It looks like something is wrong with the civet recipe (or did I mess up the module update?).

Any advice?

**Comment by permcody** [comment]
Can't scroll through all that on my phone. I'll have to look using my
laptop.

**Comment by permcody** [comment]
Submodule looks ok, it's definitely isolated to your PR though. Mike's is
building just fine. You can see the environment and the exact command we
run in the output. I don't see the problem right this second...

File: scripts/update_and_rebuild_libmesh.sh
Comment (permcody) [inline scripts/update_and_rebuild_libmesh.sh]: I'm going to guess the problem with the reason this PR didn't work has nothing at all to do with CIVET...

**Review by permcody — CHANGES_REQUESTED** [review]

File: scripts/update_and_rebuild_libmesh.sh
Comment (friedmud) [inline scripts/update_and_rebuild_libmesh.sh]: lol!

Whoops - that slipped in there. 😆

**Comment by permcody** [comment]
image

**Comment by permcody** [comment]
image

**Comment by permcody** [comment]
@friedmud - Do you want us to take this over? I'd like to get it in.

**Comment by snschune** [comment]
How operational is this patch?

**Comment by permcody** [comment]
I rebased it in #8557. It seems to work on some problems but it also fails the very simple test checked in here on Linux, not sure why.

**Comment by snschune** [comment]
I was testing this patch on a triangular mesh created with @YaqiWang mesh generator in instant. 
It fails with messages as below on multiple (or all) processors. I can share the input off-thread.  

```
*** ERROR ***
there are 210 side(s) uncovered by local side sets
```

**Comment by permcody** [comment]
Oh! mesh modifiers are going to need a complete overhaul. Many of them
aren't going to work at all in their current form.

Longer term, we may be able to apply mesh modifiers in the checkpoint
splitter if it's incorporated into MOOSE. Ideally we'd run all modifiers
before splitting the mesh. Have you tried running --mesh-only first then
running the splitter?

**Comment by snschune** [comment]
I am planning to use mesh modifiers to extrude this mesh, but I would do it _before_ in a `--mesh-only` step and then split. 

In this case, I am using a 2D triangular mesh and I do not apply any mesh modifiers.

**Comment by permcody** [comment]
Then I guess this thing is still pretty broken!

**Comment by snschune** [comment]
@permcody  I'm curious, why do you suspect I was using mesh modifiers? Maybe I am missing something here.

**Comment by permcody** [comment]
I just assumed.

**Comment by snschune** [comment]
Does anyone has a compiled version of this, maybe @friedmud or @permcody? 
I tested this on another problem and failed with the same error message. I am wondering
if I set it up correctly. I'd provide the input files off-list.

**Comment by permcody** [comment]
Fails for me too. Something is wrong with it.

**Comment by YaqiWang** [comment]
This error message is from an internal Rattlesnake sanity check. Is it possible that the ghost elements are not set up properly causing the checker to think the processor interface is boundary but no side sets covering it?

**Comment by snschune** [comment]
That is the error message I encounter, but I am not sure @permcody encounters the same problem. I've never provided him with my inputs so I assume @permcody ran the tests included in this PR with
MOOSE.

**Comment by snschune** [comment]
@friedmud @permcody  Is this currently on anyone's radar? It's sort of a biggie for us.

**Comment by permcody** [comment]
Not on my radar. Roy may get to it before I do. I have no idea what's going
on with Derek these days

**Comment by snschune** [comment]
Bummer. If I can help in any way, let me know. Any estimate of how long @roystgnr @friedmud or @permcody would need to get to this?

**Comment by friedmud** [comment]
I'm just currently swamped with stuff here.  I don't have any plans to get
to this for the next week at least.

**Comment by roystgnr** [comment]
Someone want to fill me in on the desired semantics here?  If https://github.com/libMesh/libmesh/pull/1291 is working the way it's intended, then CheckpointIO should be able to operate more seamlessly now.  You'll still want to set parallel() etc. when writing a CheckpointIO file, but when reading one in it'll autodetect whether the file is serial or parallel and will do reads (and if necessary, broadcasts) accordingly.

**Comment by permcody** [comment]
Our ideal usage would be to switch back and forth between different formats and modes with minimal changes to our input files. This PR added 

1. Run with a serial exodus input mesh in replicated mode - DONE
2. Run with a serial exodus input mesh in distributed mode - DONE
3. Run with a distributed nemesis input mesh in distributed mode - CHANGING SYNTAX
4. Run with a serial checkpoint mesh in replicated mode - NOT DONE
5. Run with a serial checkpoint mesh in distributed mode - NOT DONE
6. Run with a distributed checkpoint mesh in distributed mode - NOT DONE

I believe this PR does some of that if not all of that (if it was working properly). When we were in College Station we talked about having a Boolean in our Mesh block that tells MOOSE whether or not the mesh is "pre-split". When that option is on, we look for file names with the per processor suffix. When it's not on, we don't so it makes it easy to switch between input serial and distributed meshes. Likewise, we already have a `distribution` option which tells the framework which mode to run in.  

I'm not sure if I've fully answered your question. This PR should be a good start as I think the semantics are already in here. We just need to get it working.

**Comment by friedmud** [comment]
@permcody has this right.

**Comment by permcody** [comment]
Superseded by #8557.

**Pull request closed without merging**
